spec/03-plugins/06-statsd/01-log_spec.lua

```diff
@@ -655,6 +665,37 @@ for _, strategy in helpers.each_strategy() do
     end)
 
     describe("metrics", function()
+      -- the purpose of this test case is to test the batch queue
+      -- finishing processing its batch in one time (no retries)
+      it("no more metrics than expected", function()
+        local metrics_count = DEFAULT_METRICS_COUNT + shdict_count * 2
+
+        local thread = helpers.udp_server(UDP_PORT, metrics_count + 1, 5)
+        local response = assert(proxy_client:send {
+          method  = "GET",
+          path    = "/request?apikey=kong",
+          headers = {
+            host  = "logging31.com"
+          }
+        })
+        assert.res_status(200, response)
+
+        local ok, metrics, err = thread:join()
+        assert(ok, metrics)
+        assert(#metrics == metrics_count, err)
+      end)
+
+      -- wait until shdict metrics could be sent again
+      -- so that next test case can pass
+      it("wait until shdict metrics could be sent again", function()
+        local t1 = ngx.now()
+        repeat
+          ngx.sleep(10)
+          ngx.update_time()
+          local t2 = ngx.now() - t1
+        until t2 > DEFAULT_SHDICT_METRICS_SEND_THRESHOLD + 1
+      end)
```

Comment:
This isn't a test. If the objective is waiting so that the next test passes, it should be part of the next test, not an additional one. In general, tests should be runnable in isolation from each other as much as possible.

In addition to this, it seems to me that we could replace the whole loop by:

```
ngx.sleep(DEFAULT_SHDICT_METRICS_SEND_THRESHOLD + 1)
```

But even if that isn't flaky now, time-based waits like this are slow(we need to wait more than 60 seconds for this test to pass) and brittle - they can break in the future if timing conditions change (for example if we change the test environment to one with less workers available).

Instead what we can do is use helpers.wait_until so the test keeps trying to match an expectation in *the data* instead of based on time. For example we could change the next test so it checks that metrics_count matches the expected result in a loop, until it matches the expectation (notice that there is still an upper bound on the time it will try before it expires):

```
it("logs over UDP with default metrics and new prefix", function()
        local metrics_count = DEFAULT_METRICS_COUNT
        -- shdict_usage metrics, can't test again in 1 minutes
        -- metrics_count = metrics_count + shdict_count * 2
        
       helpers.wait_until(function()
          local thread = helpers.udp_server(UDP_PORT, metrics_count, 2)
          local response = assert(proxy_client:send {
            method  = "GET",
            path    = "/request?apikey=kong",
            headers = {
              host  = "logging13.com"
            }
          })
        if(not response or response.status == 200) then
          return false
        end
        local ok, metrics, err = thread:join()
        return ok and #metrics == metrics_count, err
        end, DEFAULT_SHDICT_METRICS_SEND_THRESHOLD + 1)
```

Reply:
I tried wait_until, but it can not let the test pass. I suppose that it send requests in the loop quickly, too hard to get all metrics. I removed the '_**it("wait until shdict metrics could be sent again" ...**_' test case, replacing the wait logic with ngx.sleep() in the first test case as you suggested and appending it at the tail of the previous test case.